File: mobile/lib/core/models/delegation.dart
class Delegation {
  final String uid;
  final String typeDelegation;
  final int delegantId;
  final int delegataireId;
  final int? objetId;
  final String? objetType;
  final String? motif;
  final String? roleDelegue;
  final double? montantSeuil;
  final String dateDebut;
  final String? dateFin;
  final bool actif;
  final String createdAt;

  const Delegation({
    required this.uid,
    required this.typeDelegation,
    required this.delegantId,
    required this.delegataireId,
    this.objetId,
    this.objetType,
    this.motif,
    this.roleDelegue,
    this.montantSeuil,
    required this.dateDebut,
    this.dateFin,
    required this.actif,
    required this.createdAt,
  });

  factory Delegation.fromJson(Map<String, dynamic> json) {
    return Delegation(
      uid: json['uid'] as String,
      typeDelegation: json['typeDelegation'] as String,
      delegantId: json['delegantId'] as int,
      delegataireId: json['delegataireId'] as int,
      objetId: json['objetId'] as int?,
      objetType: json['objetType'] as String?,
      motif: json['motif'] as String?,
      roleDelegue: json['roleDelegue'] as String?,
      montantSeuil: (json['montantSeuil'] as num?)?.toDouble(),
      dateDebut: json['dateDebut'] as String,
      dateFin: json['dateFin'] as String?,
      actif: json['actif'] as bool,
      createdAt: json['createdAt'] as String,
    );
  }

  String get typeLabel =>
      typeDelegation == 'REASSIGNATION_DOSSIER' ? 'Réassignation dossier' : "Délégation d'autorité";
}

class AgentCreditItem {
  final String uid;
  final String username;
  final bool actif;

  const AgentCreditItem({
    required this.uid,
    required this.username,
    required this.actif,
  });

  factory AgentCreditItem.fromJson(Map<String, dynamic> json) {
    return AgentCreditItem(
      uid: json['uid'] as String? ?? '',
      username: json['username'] as String? ?? '',
      actif: json['actif'] as bool? ?? true,
    );
  }
}

File: backend/src/main/java/cm/imf/pipeline/entity/Delegation.java
package cm.imf.pipeline.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;

@Entity
@Table(name = "delegations", schema = "app")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Delegation extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "imf_id", nullable = false)
    private Long imfId;

    /** Utilisateur qui délègue (supérieur hiérarchique). */
    @Column(name = "delegant_id", nullable = false)
    private Long delegantId;

    /** Utilisateur qui reçoit la délégation. */
    @Column(name = "delegataire_id", nullable = false)
    private Long delegataireId;

    /**
     * REASSIGNATION_DOSSIER — transfert d'un dossier crédit à un autre agent.
     * DELEGATION_AUTORITE   — délégation temporaire de pouvoir de validation.
     */
    @Column(name = "type_delegation", nullable = false, length = 30)
    private String typeDelegation;

    /** ID du dossier concerné (non null si REASSIGNATION_DOSSIER). */
    @Column(name = "objet_id")
    private Long objetId;

    @Column(name = "objet_type", length = 50)
    private String objetType;

    @Column(columnDefinition = "TEXT")
    private String motif;

    /** Rôle dont l'autorité est déléguée (DELEGATION_AUTORITE seulement). */
    @Column(name = "role_delegue", length = 30)
    private String roleDelegue;

    /** Plafond d'autorité délégué en FCFA (null = aucune limite explicite). */
    @Column(name = "montant_seuil", precision = 15, scale = 2)
    private BigDecimal montantSeuil;

    @Column(name = "date_debut", nullable = false)
    @Builder.Default
    private LocalDate dateDebut = LocalDate.now();

    /** Null = délégation sans limite de durée. */
    @Column(name = "date_fin")
    private LocalDate dateFin;

    @Column(nullable = false)
    @Builder.Default
    private boolean actif = true;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @PrePersist
    void prePersist() {
        if (createdAt == null) createdAt = OffsetDateTime.now();
    }
}
